### ISSUE TITLE
Async beforeAll/afterAll problems

### ISSUE THREAD

**Issue by cmidgley** [wallabyjs/public#2663]
### Issue description or question

I have been trying to track down some nasty lost resources problems, and I've finally discovered that Wallaby does not seem to be honoring the requirements of `beforeAll`/`afterAll` when `async` is being used.  Here is a simple test case:

```
describe("test", () => {
	beforeAll(async () => {
		console.log("beforeAll starting");
		await new Promise((resolve) => {
			setTimeout(() => resolve(), 1000);
		});
		console.log("beforeAll complete");
	});
	afterAll(async () => {
		console.log("afterAll starting");
		await new Promise((resolve) => {
			setTimeout(() => resolve(), 1000);
		});
		console.log("afterAll complete");
	});
	it("should be ok", () => {
		expect(3).toBe(3);    
		// expect(3).toBe(5);
	});
});
```
Gist: https://gist.github.com/cmidgley/35de597abd7ca64d9755d31ab71bafdb

For this, I am using just automatic configuration for Wallaby (no config file) and there is just the default `package.json` with a simple jest test script:

```
{
	"name": "jest-test",
	"version": "1.0.0",
	"description": "",
	"main": "index.js",
	"scripts": {
		"test": "jest"
	},
	"keywords": [],
	"author": "",
	"license": "ISC",
	"devDependencies": {
		"jest": "^26.6.3"
	}
}
```
Gist: https://gist.github.com/cmidgley/2137dbc28e7972ff8f31e28b083e8489

As you can see, I have a one second delay in `beforeAll` and again in `afterAll`.  When Wallaby starts up, these execute just as expected:

```
‍[Info]  Started Wallaby.js Core v1.0.1045
[Info]  console.log: beforeAll starting
[Info]  console.log: beforeAll complete
[Info]  console.log: afterAll starting
[Info]  console.log: afterAll complete
[Info]  console.log: Jest: PASS tests/sample.test.js
```

However, when I then edit the test file and just add spaces to it at a rate faster than the test execute, I find that `beforeAll` never completes, and `afterAll` is only called once:

```
[Info]  console.log: beforeAll starting
[Info]  console.log: beforeAll starting
[Info]  console.log: beforeAll starting
[Info]  console.log: beforeAll complete
[Info]  console.log: afterAll starting
[Info]  console.log: afterAll complete
[Info]  console.log: Jest: PASS tests/sample.test.js
```

Here is a gif of the experience:

![beforeall](https://user-images.githubusercontent.com/8160316/111192074-1095c380-858f-11eb-93aa-c19177dd9b10.gif)

In my use case, `beforeAll` creates resources including persistent database records, so calling it multiple times is a big problem, and not calling `afterAll` leaves the resources unreleased.

Am I doing something wrong here, or is this a bug?  Also, I find that the same process Id is reused, yet my process-wide members seem to have a separate instance (I've not proven this yet, so I could be wrong) - but I was wondering what Wallaby does for running these in different sandboxes?  I looked at `worker.threadId`, but that doesn't seem to change so I'm guessing that isn't being used?

Thanks for looking at this!

**Comment by cmidgley** [comment]
See #2660 - similar issue with globalSetup, thinking these may be related issues?

**Comment by smcenlly** [comment]
Thanks for providing such a nice summary in your issue response. It made it really easy for us to understand your problem.

> See #2660 - similar issue with globalSetup, thinking these may be related issues?

I've read #2660 (the `globalSetup` issue), but need to do a bit more review before responding to that one. This issue may be related but the behavior you mention above is expected.

> Wallaby does not seem to be honoring the requirements of beforeAll/afterAll when async is being used.

This is not actually a requirement of your testing framework and no guarantee is actually provided - I'll explain more below.

When Wallaby is running it may cancel a current test run after it has started and terminate it mid-way through execution. Wallaby gives your test some time to finish executing but if it takes more than a second or so, it will terminate execution. Wallaby does this to minimize CPU processing times and provide fast feedback in your editor. This is by design. Unfortunately, sometimes this may lead to odd behavior in your tests when you rely on some process state that is set in `beforeEach` or `beforeAll` and `afterEach` and `afterAll` do not execute. Your testing framework also does not offer any guarantee on executing this logic if for example you were to stop the process/test run half-way through its execution. Having said that, in Wallaby any problems relating to this behavior can be exacerbated due to what Wallaby is doing.

You have the exact same behavior (and testing framework guarantees) with `globalSetup` and `globalTeardown`. These functions are *not* guaranteed to run (e.g. `globalSetup` may run but a global error may halt your tests and `globalTeardown` is never executed). From the context of running your tests in Wallaby, you need to make sure that your code is not fragile to this behavior as it is more likely to occur than using the CLI. Because Wallaby by default reuses its worker processes, if your code does not accommodate the case where `globalTeardown` is not run, you may have problems.

You may use the [workers configuration](https://wallabyjs.com/docs/config/workers.html) to mitigate this behavior by setting the `restart` property to `true` which will tell Wallaby not to reuse its worker processes. If you do so, I expect you would notice a decrease in test run feedback performance; how much depends on your project / setup. If possible, we recommend updating your code to support multiple calls to `globalSetup` without `globalTeardown` (and the same for `before*` vs. `after*`).

**Wallaby.js workers example**
```javascript
 module.exports = function () {
   return {
     autoDetect: true,

     workers: {
       restart: true // Restart a new process for each subsequent test run
     }
   };
 };
```

Also, by default Wallaby parallelizes your test runs so depending on your setup code and what it's doing, you may also need to limit worker concurrency by setting the worker `initial` and `regular` settings both to `1`. Again, this will slow down Wallaby execution but it may be acceptable to you given your project setup logic.

---

> Also, I find that the same process Id is reused, yet my process-wide members seem to have a separate instance (I've not proven this yet, so I could be wrong) - but I was wondering what Wallaby does for running these in different sandboxes? 

I assume by process-wide members you are referring to something that is set in `globalSetup` or `globalTeardown`? Wallaby runs your `globalSetup` in its own process, then spawns different worker processes to run your tests, and then runs your `globalTeardown` in the same process that ran `globalSetup`. In short, `globalSetup` and `globalTeardown` are not expected to run in the same process when using Wallaby. Outside of Wallaby where your testing framework may use a single process to run all of your tests, this behavior may indeed be different (depending on your test runner). Wallaby also passes `process.env` variables that may have been changed in `globalSetup` to the worker processes.

> I looked at worker.threadId, but that doesn't seem to change so I'm guessing that isn't being used?

The `worker.threadId` can (and does) change from time to time depending on what is happening in your project. If you are only editing one test then you are likely to get the same worker id over and over again but if you are changing code that affects multiple test files, you may find the ids change more often. Also, if there is an error or timeout running your tests, you may find that Wallaby will terminate a worker process and create a new one.

Most projects don't have process-wide context so your kind of issue is a little atypical but your scenario is certainly supported by Wallaby and we can help get it up and running for you. If you have specific project code you can share with us (publicly or privately via hello@wallabyjs.com) we would be happy to take a look at your setup/teardown logic to help you come up with a solution.

**Comment by cmidgley** [comment]
Thank you for the detailed writeup.  While I understand your perspective, I do think there is a critical difference between Jest and Wallaby here.  When Jest fails, the process is torn down and all resources are released (like it or not) and no more tests are executed on that process.  But with Wallaby, the process lives on (assuming I don't tell it to restart processes), which means the resources remain consumed.  This is a very important distinction.  

In my case - I reserve a database instance in `beforeAll` (a schema and many tables) from a pool, make socket connections to services, start up an Express/Apollo/Nexus server, and more.  I need some trigger (even if in another `beforeAll` call, see below) to release the prior resources, else I rapidly run out of resources (especially while typing and having Wallaby set to automatic) and have to shutdown and restart Wallaby (and sometimes VS Code).

> From the context of running your tests in Wallaby, you need to make sure that your code is not fragile to this behavior as it is more likely to occur than using the CLI. 

I agree that you must be super careful with resources in Wallaby (I've spent the last 10 days focused solely on Wallaby to get to this point).  Are you saying the code needs to be capable of having it terminated at any time, even mid-method (such as when waiting on a promise), and re-entered again without any signal or message?  That's possible, though tricky when dealing with persistent resources like databases, and even trickier to do it with performance.  I think to do that there would need to be a guarantee that anytime `beforeAll` is called, that no more pending promises will be completed from a prior run, and no `afterAll` will be called to close that prior run (for that process).  Then the code could (in theory), at the start of `beforeAll`, execute a cleanup of the prior `beforeAll`.

I say "in theory" because I find that if I create a variable at module scope, and increment it each time `beforeAll` is called, it never changes between `beforeAll` executions.  My point being, I'm unclear what memory is shared from the prior `beforeAll` to the next `beforeAll` (on the same process ID).  This may be my lack of understanding how JS engines work?  For example (untested code follows... to illustrate my point):

```
...
let myGlobal = 0;
beforeAll(async () => {
  ++myGlobal;
  console.log(myGlobal);
}
```

When I run code similar to the above, I always see a value of `1` for `myGlobal`.  The obvious problem here is that if I am supposed to use `beforeAll` to clean up my resources from the prior `beforeAll` execution, I need to access the properties and resources associated with the prior run.  But in this case, I can not because they are not there.

I feel like I missing something obvious here.  How am I supposed to be able to clean up resources allocated in `beforeAll`, upon a future `beforeAll` call, when `afterAll` is not called?

Thank you very much for your help and time on this!

P.S. it would be great if the various timeouts used by Wallaby where in the `wallaby.js` configuration file.  For example, I get warnings about the sandbox being dead, yet it's just the fact that I must have some long running processes.  And you mentioned the Wallaby tears down my tests after just one second - faster than my resources will release from the third party libraries that I consume.

**Comment by ArtemGovorov** [comment]
> While I understand your perspective, I do think there is a critical difference between Jest and Wallaby here. When Jest fails, the process is torn down and all resources are released (like it or not) and no more tests are executed on that process. But with Wallaby, the process lives on (assuming I don't tell it to restart processes), which means the resources remain consumed. This is a very important distinction.


Can you please provide an example of "When Jest fails, the process is torn down and all resources are released", and the opposite in Wallaby?

Here's a example when Jest and Wallaby are consistent and yet will both not complete the `afterAll` call (making it simply unreliable for the cleanup purposes regardless of the tool):
- run Jest in watch mode, print process id in one of your tests,
- comment out the `setTimeout(() => resolve(), 1000);` line in your example afterAll function,
- save the file,
- wait for a few seconds,
- uncomment the commented line,

There'll be total 2 runs, in the same process (as you may check from the output). In the run #1, the `afterAll` function will not be completed.

**Comment by cmidgley** [comment]
Interesting for sure - thank you figuring this out and sharing!  This, plus a few hours of being unable to sleep last night (!), has helped me to understand I've had a knowledge gap surrounding process reuse on test restart.

It's feeling like when a test is restarted, and the process is reused, something such as the JavaScript VM is being restarted (or if not a VM restart, perhaps something similar)?  Meaning, the JS in-memory resources that were associated with that process are destroyed (not released, but eliminated none the less). Do you know if this a correct understanding?

If it is, then it sounds like I can trust that `beforeAll` (and `globalSetup`, etc) can be safely assumed to be fresh (no prior test resources consumed) and that I can safely use `process.pid` as a key on my persistent store to figure out how to release and reuse my external resources (in my `beforeAll`, start with cleanup of any prior usage keyed on `process.pid`).

For example (pseudocode):

```
beforeAll(... {
  // tear down prior usage of persistent store using our process ID as a key
  if (storeHas(process.pid))
    releaseStore(process.pid);

  // setup our persistent store for this test
  setupStore(process.pid)
});
```

I think this issue (how Jest reuses processes by resetting the VM) also helps to clarify #2660 for me.

Thank you again.  I think I'm very close...

**Comment by cmidgley** [comment]
I'm really sorry about this long post, and all the work I've been putting on your plate(s)... but I'm really getting stuck here and am hoping we can find a pathway that let's us use Wallaby.  

**TL;DR?** Everything works perfectly under Jest including Watch mode, but fails consistently with lost resources under Wallaby with no path that I can see to resolution (short of restart process which is painfully slow).  

### Issue one: Unable to reproduce the `afterAll` test in Jest recommended above

I did get to trying your recommendation [above](https://github.com/wallabyjs/public/issues/2663#issuecomment-799919331) to experience the failure in Jest to call `afterAll`.  I am unable to reproduce this.  I get a new process ID each time and always see `afterAll` on completion doing the test you recommend.  Further, I have switched to using `--watch`  on native Jest (no Wallaby) on my main project and after at least 50 runs I've had zero resource leaking, and I've receive the `afterAll` on every single test (including individual tests and full test suites).  Perhaps I misunderstand the options to use to reproduce this in Jest?  I'm using:

`package.json` fragment:
``` 
"test:watch": "cross-env NODE_PATH=src/ NODE_ENV=test npx --silent jest --watch",
```

### Issue two: Memory resources are not released with `{restart: false}`

I am quite confident that my `afterAll` code releases all resources, as I can run it over and over and never see any left over connections, failures to terminate, etc.  But it's clear that inside Wallaby (never in Jest) that resources are getting lost.  For example, I use `node-postgres` and it maintains a pool of sockets to the postgres server.  These resources are not released if I don't get a chance to call `node-postgres` and tell it to shut down.  I see at the postgres server console where server session count climbs each time Wallaby terminates and restarts a test, as well as later in the Wallaby console where I get this:

```
[Info]  console.error: (node:30956) MaxListenersExceededWarning: Possible EventEmitter memory leak detected. 11 uncaughtException listeners added to [process]. Use emitter.setMaxListeners() to increase limit
```

Usually not much later I have to restart Wallaby (sometimes VSCode) in order to continue working (as often as every 30 minutes if I'm doing a lot of test work).

Since I never receive a call to `afterAll` (or anything else I'm aware of, that operates in my last used memory space) when a test is prematurely terminated (to allow me to do this cleanup), and Jest/Wallaby does not appear to be tearing everything down (such as process termination would), I fail to understand how to eliminate resource leaks (such as `node-postgres` sockets) using Wallaby short of setting `workers: { restart: true }` in `wallaby.js` (which does work, but causes my tests to go from 1-3 seconds out to 12-15 seconds). Again, these issues never show up in Jest, including Jest watch mode.

### Issue three: Strange issue with process memory changing

Something really crazy is happening.  I don't yet have a way to reproduce this for you (I am working on that), but I'll describe what I'm seeing and see if perhaps you have any thoughts...

To help me track what is going on, each time `beforeAll` starts I instantiate an object that in it's constructor it creates a random number (0-1000) and never changes it and I store that object in a global variable defined my main code file.  On various debug messages, I always output the process ID as well as that random number, so I can track which process and which memory state I'm looking at.

In my `afterAll`, I tell `node-postgres` to cleanup resources.  It always returns right away when in Jest, but upon test restarts in Wallaby it sometimes (perhaps one out of 10 test restarts) does not complete during that test run execution (not sure why, and it only every happens in Wallaby).  That means `afterAll` was blocked and never completed it's promise. 

However, later when when another test runs, `node-postgres` decides to go ahead and complete that promise.  That means the promise is still alive, even though the test has been "torn down" (similar to the resources issues above, it's clear that when sharing the process resources do remain outstanding).   That's not desirable, but can be handled.

But what's really strange is that when that completion runs, it has a different ID!  It's the ID of a more recent run that started long after that run started.   I see a pattern that looks like this when a test is restarted:

```
beforeAll starting ID 33
beforeAll finished ID 33
beforeAll started ID 86
beforeAll finished ID 86
endAll starting ID 33
endAll starting ID 86
endAll finished ID 86
```

Notice that ID 33 never finished the endAll.  Several minutes later, I might make a change and the same code executes:

```
beforeAll starting ID 179
beforeAll finished ID 179
beforeAll starting ID 561
afterAll starting ID 561
afterAll finished ID 179
afterAll finished ID 561
```

Note the `afterAll finished ID 179` - there was no `afterAll starting ID 179` call.  This I believe is the completion of the prior `afterAll starting ID 33` - but now with a different ID (thereby, different memory space) which makes no sense at all (to me at least).  Again - I'll point back to my lack of understanding how the memory space is being managed - it doesn't seem as simple as just process memory, yet I never see a `worker.threadId` other than 0, so I remain unclear on how the working memory is recycled for each test run (since my variables seem to be reset to undefined at the start of the run, and yet consumed resources are not destroyed).

Now I'll admit there is a lot of code involved in my system (17K lines of my code, and around 1,000 node modules) and I fully understand if you don't trust these results (they do seem quite unbelievable to me as well, unless you have some knowledge or have an epiphany that can help explain this).  Next I plan on spending some time trying to reproduce this in a small test bed.

**Comment by cmidgley** [comment]
## You might want to ignore the above as I **think** I may understand this issue now

I've been researching Jest and it's now clear it uses node.vm to sandbox the code running in the process.  I'm finding that in Jest native (watch mode or not) it never executes more than one sandbox at the same time within a test (at least in my environment), but within Wallaby it does.  My problems occur exclusively when this occurs.

I wrote a bunch of small tests and I see that simple native resources, even in ES6 imported modules, all get released when Wallaby runs the test (thanks to the sandbox).  Yet I see resources in my complex app not always release (again, when multiple vm's run on a single process).  So my best guess now is that one or more modules I depend upon must somehow consume system level resources that must be cleaned up or they are lost (meaning, they are not part of the sandbox).  If I'm correct, this means that without the `afterAll` (or Wallaby provided termination API) I am unable to keep those resources from being lost (short of using a new process on each run, which is super slow).  I wouldn't be surprised if the failure with the odd callbacks are from the same cause (if you read Issue Three, above) - something outside of the context of the node vm initiating the callback event (just guessing here) and somehow its getting attached to the now-current VM in the process?

#### Proposed solution

I really need some callback to clean up my resources before Wallaby starts a new test run.  I see that it is possible to publish into the sandbox globals, which I was thinking might be able to be an optional hook for Wallaby to call before tearing down the test.  I don't know if Wallaby already has this level of access (adjusting globals) directly in Jest, but it appears it can via a custom test runner.  This would not break any existing use, but would allow a teardown to fix this issue.  If the function is defined, it would be the responsibility to the handler to execute a resolve callback when the teardown is complete (to allow for async teardowns) so Wallaby could then proceed to kill the Jest sandbox.  

Another option you might prefer, which would be more agnostic to the test platform, is to have Wallaby provide an export that  can be consumed by the test author. 

The interface could be just a simple callback, or perhaps even an object that includes an `EventEmitter` so you have the ability to send other messages (for future extensibility).  The following is an example of how it might be consumed once it was funneled down into the globals of the test:

```
wallaby.on('end', (resolve) => {
  void (async () => {
    await closeService1();
    await closeService2();
    resolve();
  })();
}
```

This could potentially also provide a conduit to add other features in the future, such as for a test to inform wallaby of a long running test (setting an expected duration), or to even determine if Wallaby is running in the test (if `wallaby` is undefined, it is not running, rather than `process.env.WALLABY_PRODUCTION`). 

Thanks for all your help, and for your consideration of supporting this capability.

**Comment by smcenlly** [comment]
> Next I plan on spending some time trying to reproduce this in a small test bed.

If you had trouble reproducing on a smaller project, could you try cloning your project and then start bisecting it until you have a smaller reproduceable sample? It would be ideal if you could get down to a single test file and then start removing tests until you only have 1-2 tests with the problem. At that point you may have a better idea of how to create a small test bed that you can share with us.

> I'm finding that in Jest native (watch mode or not) it never executes more than one sandbox at the same time within a test (at least in my environment), but within Wallaby it does. My problems occur exclusively when this occurs.

This makes sense. Wallaby can at times execute multiple test runs in the same process (in different sandboxes) at the same time. This is intentional but obviously is causing problems with your project that we need to help you resolve.

> Proposed solution

Regarding your proposed solution, we'd really like to fully understand what's going on first before we look at how to solve the problem.

---

Is it possible to move your `beforeAll` / `afterAll` code into `globalSetup` / `globalTeardown`? I suspect this would be more reliable for you and may solve your problem.

Note sure if you are aware but we also have a [discord channel](http://wallabyjs.com/chat). If you want to chat about this in real-time, you may also find us there.

**Comment by cmidgley** [comment]
I will try to reduce the code base, though it will mean writing custom tests as what will be left won't run any existing tests. 
But I think that is the right next thing to do, as my gut says it's one of two modules, both of which are somewhat independent from the core system.  I will start working on that next (tomorrow) and see if I get anywhere interesting.

Switching to `globalSetup` / `globalTeardown` won't really work for me, as all my tests are built around the environment being setup (and custom) to each test suite.  Additionally, per #2660, I saw `globalSetup` only on the first worker in the first process, and not on the other processes, which would make it impossible to use if that is the case.  I would need to go back and look at that again, as I have a far better understanding of workers/VM/processes than I did before.

Maybe the biggest thing I don't understand, and perhaps you have insight on, is how can a test (more specifically, a VM) lose resources after a sandbox is torn down?  Technically, what could cause the loss of resources (such as event listeners overflowing and postgres connection handles being lost)?  My testing with Wallaby in a small test bed has shown that outstanding promises are terminated, shared Event Listeners are released, ES6 imports released their resources, sockets are released, etc.  I've tried everything I can think of to lose resources and have failed.

The most common reason I've found online for why this type of behavior are external resources not inside the VM space (such as modules that consume C++) that never received their call to be released.  Since JavaScript does not have destructors, the external code is never notified, and so the resources remain held until the process terminates.  I know that if I do call to release the resources (using `afterAll` w/Jest), they release efficiently and consistently (Jest and Jest Watch, hundreds of tests back-to-back with no problems; under Wallaby my tests fail within a minute).

On a different, I just tried to see if I could use the `wallaby.js` `setup()` function to set a global variable that would be available in the test files.  I was successful doing this in a `jest.setup.js` file, but when I adjust `global` in `wallaby.js` it does not seem to appear in the test itself.  The [Wallaby Setup page](https://wallabyjs.com/docs/config/bootstrap.html) says:

> However, because the setup function is executed in the context of the test runner, you can use its environment to initialize it the way you need.

> For example, you can access the window object if you are running browser tests or use the global object in case of node.js tests (or the require modules you need).

Which I took to mean I could modify global and see it in my test.  In `wallaby.js` I do see the `global` variable, and I can add a new member to it (and see other properties on it as well).  But in my test code (or `jest.setup.js`, I tried both) I cannot access it as a global variable, as `global.variableName` or as `wallaby.variableName` (that was a wild guess, but thought I'd try).  Any hints how to set a global variable in `wallaby.js` and access it in my test?

As always thanks for your help.  I should go check out the discord channel - maybe tomorrow I'll do that.  I'm GMT-4 ... and my day is typically 5AM - 9PM.  Guessing our discord hours are limited...?

**Comment by smcenlly** [comment]
> Switching to globalSetup / globalTeardown won't really work for me, as all my tests are built around the environment being setup (and custom) to each test suite. Additionally, per #2660, I saw globalSetup only on the first worker in the first process, and not on the other processes, which would make it impossible to use if that is the case. I would need to go back and look at that again, as I have a far better understanding of workers/VM/processes than I did before.

Yep - makes sense.

> Maybe the biggest thing I don't understand, and perhaps you have insight on, is how can a test (more specifically, a VM) lose resources after a sandbox is torn down? Technically, what could cause the loss of resources (such as event listeners overflowing and postgres connection handles being lost)? My testing with Wallaby in a small test bed has shown that outstanding promises are terminated, shared Event Listeners are released, ES6 imports released their resources, sockets are released, etc. I've tried everything I can think of to lose resources and have failed.

We were discussing this as a team earlier and can think of some potential race conditions that might cause a problem, especially if you are using process.pid as some sort of key, but it seems unlikely.

> The most common reason I've found online for why this type of behavior are external resources not inside the VM space (such as modules that consume C++) that never received their call to be released. Since JavaScript does not have destructors, the external code is never notified, and so the resources remain held until the process terminates. I know that if I do call to release the resources (using afterAll w/Jest), they release efficiently and consistently (Jest and Jest Watch, hundreds of tests back-to-back with no problems; under Wallaby my tests fail within a minute).

Jest sandboxing might have something to do with your problem but we've not seen it before.

> Which I took to mean I could modify global and see it in my test. In wallaby.js I do see the global variable, and I can add a new member to it (and see other properties on it as well). But in my test code (or jest.setup.js, I tried both) I cannot access it as a global variable, as global.variableName or as wallaby.variableName (that was a wild guess, but thought I'd try). Any hints how to set a global variable in wallaby.js and access it in my test?

You can modify a process.env variable (string) but not pass an object. When jest is running your tests it takes over globals, creating a dummy object and resetting it for each test run.

> I'm GMT-4 ... 

We're GMT+10 so not super friendly cross over. We're usually online from 7am our time, which is 5pm your time.

---

I'll do some experimenting with how we might be able to share context between test runs in a given worker process but as per my previous comments, am a little worried about implementing a solution without fully understanding the cause of the problem (and it may not help).

**Comment by smcenlly** [comment]
We've spent the afternoon doing more analysis and work on your issue and have decided to augment the jest `global` object when running in a Wallaby context.

In your tests, when Wallaby is running, you may now access `global.nodeGlobal` which is the node process global, not the santized global that is used by jest.

This global can be modified and will remain modified between test runs. This could for example allow you to set a single database instance for each worker process instance.

We also worked out what might be happening with your existing code and why it's not working in Wallaby on key changes. When Wallaby recycles its worker processes, it is clearing orphaned timeouts and intervals so if your node modules are using `setTimeout` or `setInterval` as part of their clean up processes then this might be the cause of your problem. At any rate, you should now be able to use the `global.nodeGlobal` to enhance your cleanup logic.

We're still keen to see your code if you get a chance to package it up for us but given the time you've already invested and the solution above, we completely understand if you don't need/want to do this.

**Comment by cmidgley** [comment]
Thanks - I'll take a look at `global.nodeGlobal` later today.

Thinking about the issue this morning, what I still don't understand is the statement that `afterAll` (or at least some tear down) isn't guaranteed in an environment where processes are reused, and yet the test author owns cleaning up resources.  Those two seem at total odds - physically impossible.  As we've discussed, `beforeAll` is called in a VM and it can be destroyed without `afterAll` being called, which implies that all resources must be cleaned up by the VM destruction without any assistance from my code.  If my code is never called, how could I ever clean up anything?

Of course that is all based on the assumption that resources are VM centric.  The other side of this are persistent resources, like databases.  See [this blog](https://walrus.ai/blog/2020/04/testing-database-interactions-with-jest/) for a writeup of somebody doing Jest testing on a database (similar concept to me, but a much more simplistic environment).  Their tests will never work in Wallaby, as they rely upon `afterAll` working.  In my case, I got around this issue by writing a custom code for Wallaby that uses a persistent "pool" of database tables, dynamically creating, truncating, and garbage collecting (when they are lost due to no `afterAll`) the tables in my `beforeAll`.  It was a lot of work just to get Wallaby able to work with a database, work that was unnecessary in Jest.

All in all, it really feels like the best answer is to call `afterAll` as exists tests  are more likely to work.  If not, then provide a Wallaby specific VM teardown handler that a test author can hook (though in the end, most authors I'll bet would just call their `afterAll`, or at least a shared function, from that hook).   

I'm going to start working now on breaking apart my test to see if I can reproduce the problem in a way that can be shared with you.  Fingers crossed!

**Comment by cmidgley** [comment]
First, thank you for adding `global.nodeGlobal` to Wallaby - I have tested it and it works great, including the ability to do callbacks from the VM into the runner space.  This may end up being how we resolve this problem, but let's see as I have learned more.

For grins, I commented out all my resource cleanup in `afterAll`.  What I found is that nearly all my resources automatically release, except my postges sessions.  Those remain until I restart Wallaby

**Comment by cmidgley** [comment]
First, thank you for adding `global.nodeGlobal` to Wallaby - I have tested it and it works great, including the ability to do callbacks from the VM into the runner space.  This may end up being how we resolve this problem, but read on first...!

For grins, I commented out all my resource cleanup in `afterAll`.  What I found is that nearly all my resources automatically release, except my postges sessions.  Those remain until I restart Wallaby if I don't get a change to request to terminate them.  So I built a test for this (in this [GitHub repo](https://github.com/cmidgley/wallaby-postgres-resources)) - just setting up the `pg` module (aka, `node-postgres`) and indeed it is easy to reproduce.  

If you want to try it, you will need to install a postgres server, add a table called "testing" with any password you want, edit the code so that password is used, and then run it according to the comments in the code.  You may want to get some tool to help you monitor the session handles - again, see the code for comments on recommendations.  

But the pattern is simple and as I suspected: If I allocate a postgres handle, and don't specifically free it, when the vm is torn down the resource is lost.  In the test repo, if you let each test run complete, no session handles are lost.  But if you interrupt each test run (there is a one second sleep in the test method), you lose handles each and every test run.

I do not think this is the only example of this in the modules I am dependent upon - I have seen other conditions such as Event Listeners being over consumed after a long period of testing, and that failure has not been seen in this test example.

I'll be very interested to hear your thoughts on this.  My take is that while having a guarantee around `afterAll` (except when a process is destroyed) would be easiest for test writers, the better solution might be a pair of methods in `wallaby.js` that do basically the same (a `beforeAll` and `afterAll`) that can be hooked and then use the new `nodeGlobals` to communicate that from the runner space to the vm and let the test author worry about resource management only when it's required.  THis would be less disruptive to your installed base and would maintain your performance benefit for the majority of use cases that do not require `afterAll` to consume cycles.

**Comment by cmidgley** [comment]
I just realized I somehow closed this ... reopening!

**Comment by smcenlly** [comment]
> My take is that while having a guarantee around afterAll (except when a process is destroyed) would be easiest for test writers, the better solution might be a pair of methods in wallaby.js that do basically the same (a beforeAll and afterAll) that can be hooked and then use the new nodeGlobals to communicate that from the runner space to the vm and let the test author worry about resource management only when it's required.

I agree having a guarantee around afterAll would be easiest but it would destroy Wallaby's performance and in many cases may be unnecessary (obviously depends on the resource - definitely a problem for posgress). We also agree re your solution (something we had discussed yesterday).

In fact, Wallaby already has the hooks that you describe. The example below shows how you might utilize them:

**wallaby.js config file**
```javascript
module.exports = ({
  autoDetect: true,

  // Initialize connections array
  setup: function () {
    global.pgConnections = [];    
  },

  // Note: The teardown function will only be invoked if the setup was invoked.
  teardown: function () {
    global.pgConnections.forEach(connection => {
      // discount/cleanup logic
    });
  }
})
```


**sample.spec.js**
```javascript
describe("my tests", () => {
  var databaseConnection;
  beforeAll(() => {
    databaseConnection = getConnection();
    if (global.nodeGlobal) {
      global.nodeGlobal.pgConnections.push(databaseConnection);
    }
  });

  it("Some database test", () => {
    // code
  });

  afterAll(() => {
    if (global.nodeGlobal) {
      global.nodeGlobal.pgConnections = global.nodeGlobal.pgConnections.filter(
        (connection) => connection !== databaseConnection
      );
    }
  });
});
```

**Comment by cmidgley** [comment]
Great idea, and it hopefully will be a key part to getting us there ...  but no luck yet.   I plumbed in the change to my main code (abstracting all resources into a common class, and letting the resource destruction get managed by either `afterAll()` or via a callback from `teardown()`).  As before, I've made the teardown logic safe for partial instantiated construction so the teardown can occur even when the setup is terminated early.  However, I continue to drop the same resources on the floor.

The root of the problem appears to be that `teardown()` doesn't work with a returned Promise.  Without that, code inside `teardown` cannot stall the destruction of the VM.  See the same [github repo](https://github.com/cmidgley/wallaby-postgres-resources), now updated to no longer require Postgres (so easy to run with no other setup) and using Wallaby `setup()` and `teardown()` as recommended [above](https://github.com/wallabyjs/public/issues/2663#issuecomment-801511376).

If you run this sequentially (one at a time) it works as expected - you will see a uniqueId printed for each phase of execution and it follows the entire course of execution.  If you then cause the test to run back-to-back, you will see a different path, where a test starts, teardown starts, and then the remaining resources are not released.  This is a little confusing as Jest appears to be capturing the `console.log` output and since it is being torn down, it never sends it to the console, but you can see see the messages that occur during teardown as they are outside of Jest.

This is telling me that when `teardown()` returns, it shuts down the VM and moves on, even though I have outstanding promises that require completion to remove the resources.  Worth noting is that I originally wrote this test using Postgres, with no timers (I switched to timers to remove the dependency on Postgres), and was able to experience the same failure - so it isn't just timers that have this problem.  Also, if you remove the Promise in `teardown()` and just directly invoke the test's `teardown` you get the same result - which would be expected as nobody is awaiting their completion and the VM is destroyed before the resources are released.

I think this means to get this to work, `teardown()` will need to see if the return value is a Promise, and if so await it until completion.  Likely that is a behavior `setup()` should have as well?  At least with a Promise-aware `teardown` I should peel off the next layer of this onion and see what lies beneath!

**Comment by smcenlly** [comment]
> it shuts down the VM and moves on

The problem isn't the VM, it's Wallaby cleaning up outstanding timers/etc. (the same as before).

> I think this means to get this to work, teardown() will need to see if the return value is a Promise, and if so await it until completion. Likely that is a behavior setup() should have as well? At least with a Promise-aware teardown I should peel off the next layer of this onion and see what lies beneath!

The [setup method](https://wallabyjs.com/docs/config/bootstrap.html#setup-function) already supports working with delays and has a `delayStart()` function that you will be able to use to clean up resources before subsequent runs.

Changing the `teardown` method to support something similar would require some major architectural changes that we're not especially keen to undertake if we can get things working with `delayStart()`. The sample below shows what we have in mind:

**wallaby.js config file**
```javascript
module.exports = ({
  autoDetect: true,
  // Initialize connections array
  setup: function (wallaby) {
  	if (global.pgConnections && global.pgConnections.length) {
  		wallaby.delayStart();
  		Promise.all(global.pgConnections.map(c => c.close())).then(() => wallaby.start());
  		global.pgConnections = [];
  	}
  }
})
```
**sample.spec.js**
```javascript
describe(“my tests”, () => {
  var databaseConnection;
  beforeAll(() => {
    databaseConnection = getConnection();
    if (global.nodeGlobal) {
      global.nodeGlobal.pgConnections.push(databaseConnection);
    }
  });
  ...
  afterAll(() => {
    if (global.nodeGlobal) {
      global.nodeGlobal.pgConnections = global.nodeGlobal.pgConnections.filter(
        (connection) => connection !== databaseConnection
      );
    }
  });
});
```

This should be 100% reliable for you when Wallaby cancels a test run. For successful test runs, `afterAll` will execute and clean up immediately so you should never have any resources that haven't been cleaned up properly.

**Comment by cmidgley** [comment]
I made an important change to how I trace what the code is doing.  I switched from console.log to a custom logger that writes everything found to a JSON file, and flushes it to disk constantly (and I use a JSON log viewer, LogViewPlus (Windows only), which makes it really easy to filter/view/sort the log).  This helped solve the issue that Jest hides console messages when tests are torn down.  This was a critical step, as now I can really see what's happening in great detail (and I understand how Wallaby behaves much better now).

### What I learned

1) `wallaby.js` `setup()` is being called multiple times, even while a `delayStart` is pending.  The tests themselves do not start, but `teardown` and then `setup` happen over and over.  This meant the code in `setup` needed to handle the reentrant nature of that call, which  I handle by checking if the teardown has already started (via a state variable) and ignore the call.  While I'm past this, it might be something you want to consider in your future plans (Wallaby could do the state check prior to calling start, and avoid this issue for future users).

2) I found that `beforeAll` always completes full execution (at least in every test I've done so far, including adding long timer delays to try to create an opportunity for teardown).  That's great, but the tests themselves, and `afterAll` get partially torn down at unpredictable times and don't always complete. It has become clear that while timers get destroyed by Wallaby, promises continue execution long after the test, sometimes after many executions of the same test have been started.   To address this, I changed my code so that `afterAll` is ignored when running inside Wallaby, and only trigger the resource teardown during the next wallaby `setup` operation (where the teardown is always completed, thanks to the `delayStart` and `start` methods)

These changes have gotten me much closer - I'm often able to work for several hours before having to restart Wallaby.

### Known problems

1) Code that allocates resources inside a test can still be lost, as the test can be terminated deep inside modules outside my control.  For example, a module might do some resource allocation, use a timer (or whatever it is that Wallaby tears down), Wallaby triggers a teardown, and the code never completes.  Since the module never had a chance to return a handle to me, I'm unable to manage their release.

2) When Wallaby first runs my tests, it runs them across several processes.  During live testing (when tests are torn down and restarted rapidly), only one process appears to be used.  This means that I don't see another `setup` call in the other processes, which means I never get a chance to release the resources.   In case it matters, I have `workers: { initial: 4, regular: 4 }`).    

   I currently have 17 test suites that run across 4 processes averaging 4-5 tests per process.  That results in 3 processes * 4.5 tests = 13-14 test suites of lost resources.  I estimate the project I will have in excess of 100 test suites within the year, which will result in 75 tests with lost resources (around 20 heavy weight resources per test suite, or 1,500 total).  For illustration, here is the Postgres server sessions after the initial tests are executed and upon live testing afterwards:

   ![image](https://user-images.githubusercontent.com/8160316/111877252-d191b480-8978-11eb-9d0b-53bab438191f.png)

   The yellow highlight is when I restarted the Wallaby test run, and the red circle is when I caused a single test to run.   As you can see, the Postgres session resources are consumed and do not release (since I depend on another `setup` to execute on that process).  I tried `globalSetup` and `globalTeardown` again, but they do not help as they are called only on the first process and not with timing that is useful.

While I'm not out of the woods, I'm clearly closing in.  I am now often able to run for an hour or two, but eventually (sometimes sooner than later) I have to restart Wallaby (which results in all resources being released).  

### Help needed

1) Could Wallaby be extended to offer a config option to stop Wallaby from killing timers (and whatever else it kills).  I can totally see why you implemented that, but since it is only a partial teardown of the code (promises continue execution) it results in a highly unpredictable and uncontrollable environment for complex heavy-weight async resource applications.  Would it be reasonable to request a feature that I can set in `wallaby.js` such as `disableTimerTeardown: true` (or whatever name is descriptive) to disable that capability?  Or perhaps the ability to set how long Wallaby waits before doing the teardown (and I can set it to large number)?

2) Can you help me find an approach that let's me tear down resources attached to a process that is done running code (problem 2 above) rather having them consumed for the entire life of Wallaby?  Perhaps a round of `setup` and `teardown` when the initial test run is complete, once for each process?  Or a new API that can be called from each process when the initial test run is fully complete?  Another approach might be to find a way to know if the initial run is executing, where teardowns do not appear to happen, and I can use the real `afterAll` for those and then avoid using it during the live testing?

Really appreciate all the help.  Feels like we are getting close!

**Comment by smcenlly** [comment]
We've released a new version of Wallaby core, `v1.0.1052` that adds support for returning a `Promise` from the teardown function. You can also now do the same thing for the `setup` function (return a `Promise` instead of using `delayStart()` and `start()`.

For example:

```javascript
module.exports = function () {
  return {
    autoDetect: true,
    setup: function (wallaby) {
      // required for teardown to work, even if you don't have any setup logic.
    },

    teardown: function (wallaby) {
      // if you have teardown logic that is asynchronous, you can return
      // a `Promise` for Wallaby to wait for before starting a subsequent
      // test run
      return require('externalDepdency').teardownAsync();
    },

    globalSetupTeardownTimeout: 5000
  };
};
```

For asynchronous `teardown` functions, by default, Wallaby will terminate the worker
process and start a new worker process if the teardown logic has not been completed
within `5000ms`. This setting can be overridden by setting  `globalSetupTeardownTimeout` (milliseconds).

---

We believe that the new `teardown` behavior (allowing you to return a promise) will allow you to resolve your issue so I'll close this issue. If you're having problems or need further assistance related to this issue, please reply and we'll re-open.

**Comment by cmidgley** [comment]
Fantastic addition - much easier to use!  The great news is that I'm now seeing all `setup` and `teardown` operations running to completion.  However, the tests themselves are starting and never finishing (they are getting torn down) and if they have resources consumed they get lost forever.  This is what I was worried about in my [prior comment's issue #1](https://github.com/wallabyjs/public/issues/2663#issuecomment-803454698), but now that the other resources seem to be cleaned up it's much clearer now when reviewing the logs.  

I'm wondering if a parameter could be added that's similar to `globalSetupTeardownTimeout`, but focused on test teardown (such as `globalTestTeardownTimeout`)?  Or if you don't have a timer on that, perhaps just `globalDisableTestTeardown`?  It would need to wait until the test completes of course.

A nice, but not required, feature would be to allow the developer of the test to know if a teardown was requested, and do so in a clean manor.  For example, a `nodeGlobal.teardownRequested` boolean that could be checked in a running test, and when `true` it could ensure resources are secure and stop right away.

Also - and perhaps once this is cleaned up we are better moving this to a new issue - but I often see the first test (in a series of rapidly terminated tests) going to full completion, and the following tests early terminated.  The results of the test in the IDE are represented as of that (now obsolete) first test - for example, `//?` or `Show Value` will show values from that first test run and not the latest edited changes.  Simply doing a `Run Line Tests` brings up the correct results.  Thought I'd mention it here, as I'm not 100% sure it isn't related to this new async capability.  This issue may go away with the above suggestions (since tests would run to completion in my usage), but it does hint at a potential issue with Wallaby test running.

**Comment by smcenlly** [comment]
Thanks for your patience waiting to get back to you.

> However, the tests themselves are starting and never finishing

I think what's happening in this case is that your teardown code is finishing while some of your `test` or `afterAll` functions are still being executed. I would expect this and think we have a fix for you (see solution below).

> I'm wondering if a parameter could be added that's similar to globalSetupTeardownTimeout, but focused on test teardown (such as globalTestTeardownTimeout)? Or if you don't have a timer on that, perhaps just globalDisableTestTeardown? It would need to wait until the test completes of course.

This is actually a little trickier for us than it may at first appear because we are patching parts of testing frameworks, etc. to stop tests from being processed when a subsequent test run is scheduled and we'd have to rework that logic across the various frameworks we support. The solution below is a little hacky but should give you the behavior that you want.

> A nice, but not required, feature would be to allow the developer of the test to know if a teardown was requested, and do so in a clean manor. For example, a nodeGlobal.teardownRequested boolean that could be checked in a running test, and when true it could ensure resources are secure and stop right away.

You can actually do this yourself if you like in your `wallaby.teardown` function (example below) but I don't think you will need to do anything like this with our solution.

> Also - and perhaps once this is cleaned up we are better moving this to a new issue - but I often see the first test (in a series of rapidly terminated tests) going to full completion, and the following tests early terminated. The results of the test in the IDE are represented as of that (now obsolete) first test - for example, //? or Show Value will show values from that first test run and not the latest edited changes. Simply doing a Run Line Tests brings up the correct results. Thought I'd mention it here, as I'm not 100% sure it isn't related to this new async capability. This issue may go away with the above suggestions (since tests would run to completion in my usage), but it does hint at a potential issue with Wallaby test running.

Let's see if you still have the same problem after making the changes suggested in solution below. If so, please create a new issue for us.

### Solution

Our solution is a little hacky, but it should work for you. The idea is to create a promise and callback in your Wallaby setup function, call in in your test file's `afterAll` method, and then wait for that promise in your Wallaby teardown method. See examples below:

**wallaby.js**
```javascript
module.exports = ({
  autoDetect: true,
  files: [
    '!**/**.log'
  ],
  setup: function() {
    global.testRunDone = new Promise((resolve) => {
      global.canCompleteTestRun = resolve;
    });
  },
  teardown: function() {
    // Can be accessed from tests with `global.nodeGlobal.teardownRequested`.
    global.teardownRequested = true;

    return global.testRunDone;
  }
});
```

**sample.spec.js (test file)**
```javascript
const path = require('path');
const fs = require('fs');

function waitOneSecond() {
  return new Promise(resolve => {
    setTimeout(resolve, 500);
  });
}

describe('my tests', () => {
  const logFile = path.join(path.dirname(__dirname), 'output.log');

  beforeAll(async () => { 
    fs.appendFileSync(logFile, 'beforeAll:start\n');
    await waitOneSecond();
    fs.appendFileSync(logFile, 'beforeAll:end\n');
  });

  it('test1', async () => {
    fs.appendFileSync(logFile, 'test1:start\n');
    await waitOneSecond();
    fs.appendFileSync(logFile, 'test1:end\n');
  });

  it('test2', async () => {
    fs.appendFileSync(logFile, 'test2:start\n');
    await waitOneSecond();
    fs.appendFileSync(logFile, 'test2:end\n');
  });

  afterAll(async () => {
    fs.appendFileSync(logFile, 'afterAll:start\n');
    await waitOneSecond();
    fs.appendFileSync(logFile, 'afterAll:end\n');
    global.nodeGlobal && global.nodeGlobal.canCompleteTestRun();
  });
});
```

With this code, Wallaby will not continue executing on teardown until `afterAll` has been executed. This also means you can keep all teardown logic in your test file. There is one caveat, which is that every one of your test files will need to call `global.nodeGlobal.canCompleteTestRun();` to let Wallaby know that teardown is complete.

We've done some testing and it appears that beforeAll, afterAll are always guaranteed to execute with this configuration. There are times where not all tests will execute, but if a test starts execution, it is guaranteed to finish.

**Comment by cmidgley** [comment]
Awesome - it's working and all resources are releasing consistently.  And it appears the multiple/incorrect test run issue has gone away.

There was one more change I had to make to get it to work.  Sometimes Wallaby calls `setup()` and then `teardown()` without any tests in between.  This would cause the promise to hold until the teardown timer expired.  To fix this, I simply added a boolean in `nodeGlobal` called `testDidStart` and I set it to true inside `beforeAll`.  Then in the `teardown()` I check if `testDidStart` is false and skip waiting for the promise resolution.    

And now here is my session graph from PostGres:

![image](https://user-images.githubusercontent.com/8160316/112139200-e18fdb00-8ba8-11eb-8e96-3fdbbf8fa94f.png)

THANK YOU SO MUCH FOR YOUR AMAZING HELP!  (and kickass product)
